src/Orleans/Utils/Factory.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace Orleans
+{
+    /// <summary>
+    /// Creates an instance of <typeparamref name="TInstance"/>.
+    /// </summary>
+    /// <typeparam name="TInstance"></typeparam>
+    /// <returns></returns>
+    public delegate TInstance Factory<out TInstance>();
```

Comment:
The only issue with the suggested example, is that it doesn't support passing parameters for creating, as we need in this PR. Sure, we could in theory have `IFactory<out TInstance, in TParam1, in TParam2>` "overloads" so that the `Create` can take parameters, but at that point I believe we are just using types to emulate its functional counterpart. 
The functional approach also covers each of the items you listed. In fact if you look at `FactoryUtilities` many of those items are currently implemented, and the constructed object is being built using the container to fulfill the dependencies, but you could also do composition even easier than how you do it with the decorator pattern.
Won't push back if the class approach that's the consensus, but my preference goes to the functional approach in that case, since it's easy to map `Func<T>` to `Factory<T>` with less ceremony, and one is implicitly castable to the other if the end-user cares about that.
BTW, having a single interface and `Create` take `params object[]` is out of the question, as it doesn't make the requirements explicit.